mypy/checker.py

```diff
@@ -572,13 +572,27 @@ def is_implicit_any(t: Type) -> bool:
                     if (isinstance(defn, FuncDef) and ref_type is not None and i == 0
                             and not defn.is_static
                             and typ.arg_kinds[0] not in [nodes.ARG_STAR, nodes.ARG_STAR2]):
-                        if defn.is_class or defn.name() in ('__new__', '__init_subclass__'):
+                        isclass = defn.is_class or defn.name() in ('__new__', '__init_subclass__')
+                        if isclass:
                             ref_type = mypy.types.TypeType(ref_type)
                         erased = erase_to_bound(arg_type)
                         if not is_subtype_ignoring_tvars(ref_type, erased):
-                            self.fail("The erased type of self '{}' "
-                                      "is not a supertype of its class '{}'"
-                                      .format(erased, ref_type), defn)
+                            note = None
+                            if typ.arg_names[i] in ['self', 'cls']:
+                                if erased != arg_type or isclass:
```

Comment:
Types don't define `__eq__`, so it's better not to use `!=` here. Either use `is not` (but then we should document that `erase_to_bound` preserves the identity of the argument under certain conditions) or `is_same_type`.